text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+  trait as the name of the macro is from the perspective of the language's
+  static semantics entirely coincidental.
+
+  However, for users who aren't aware of this, it may seem strange that
+  `SmartDefault` should derive for the `Default` trait.
+
+### `#[derive(new)]`
+
+[`derive-new`]: https://crates.io/crates/derive-new
+
+The [`derive-new`] crate provides the `#[derive(new)]` custom derive macro.
+Unlike the two previous procedural macro crates, `derive-new` does not
+provide implementations of `Default`. Rather, the macro facilitates the
+generation of `MyType::new` constructors.
+
+For example, you may write:
+
+```rust
+#[derive(new)]
+struct Foo {
+    x: bool,
+    #[new(value = "42")]
+    y: i32,
+    #[new(default)]
+    z: Vec<String>,
+}
+
+Foo::new(true);
+
+#[derive(new)]
+enum Enum {
+    FirstVariant,
+    SecondVariant(bool, #[new(default)] u8),
+    ThirdVariant { x: i32, #[new(value = "vec![1]")] y: Vec<u8> }
+}
+
+Enum::new_first_variant();
+Enum::new_second_variant(true);
+Enum::new_third_variant(42);
+```
+
+Notice how `#[new(value = "vec![1]")`, `#[new(value = "42")]`,
+and `#[new(default)]` are used to provide values that are then omitted
+from the respective constructor functions that are generated.
+
+If you transcribe the above snippet as much as possible to the system proposed
+in this RFC, you would get:
+
+```rust
+struct Foo {
+    x: bool,
+    y: i32 = 42,
+    z: Vec<String> = <_>::default(),
+    //               --------------
+    //               note: assuming some `impl const Default { .. }` mechanism.
+}
+
+Foo { x: true };
+
+enum Enum {
+    FirstVariant,
+    SecondVariant(bool, u8), // See future possibilities.
+    ThirdVariant { x: i32, y: Vec<u8> = vec![1] }
+}
+
+Enum::FirstVariant;
+Enum::SecondVariant(true, 0);
+Enum::ThirdVariant { x: 42 };
+```
+
+Relative to `#[derive(new)]`, the main benefits are:
+
++ No wrapping code in strings, as noted in previous sections.
++ The defaults used can be mixed and matches; it works to request all defaults
+  or just some of them.
+
+The constructor functions `new_first_variant(..)` are not provided for you.
+However, it should be possible to tweak `#[derive(new)]` to interact with
+this RFC so that constructor functions are regained if so desired.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## `#[non_exhaustive]`
+1. What is the right interaction wrt. `#[non_exhaustive]`?
+
+   In particular, if given the following definition:
+
+   ```rust
+   #[non_exhaustive]
+   pub struct Config {
+       pub height: u32,
+       pub width: u32,
+   }
+   ```
+
+   it could be possible to construct a `Config` like so, if the construction of
+   types without default field values is allowed (to support semver changes):
+
+   ```rust
+   let config = Config { width: 640, height: 480, .. };
+   ```
+
+   then adding a field to `Config` can only happen iff that field
```

Comment:
I distinctly remember someone taking issue with "iff" being used rather than writing it out.